src/mlir/consumer.h

```diff
@@ -13,15 +13,114 @@
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/ADT/StringRef.h"
 
+#include <stack>
 #include <string>
 
 namespace mlir::verona
 {
-  /// Field offset maps between field names, types and their relative position.
-  struct FieldOffset
+  /**
+   * Class information
+   *
+   * Keeps information on classes, and their respective fields (types, names).
+   * This class is used for both keeping the context while building classes as
+   * well as querying field types by name later in code generation.
+   *
+   * FIXME: Verona may play with field offset (ex. by size order), which needs
+   * to be done *before* we access any field from code, ie. right after
+   * declaration is finished, so the order in which they appear in the vector
+   * can change but only once. MLIR doesn't allow to change the fields of a
+   * declared structure, so this needs to be done before setting the
+   * StructureType.
+   */
+  class ClassInfo
   {
-    SmallVector<llvm::StringRef> fields;
-    SmallVector<Type> types;
+    /// Bind name and type for fields. Position is given by its vector offset.
+    struct Field
+    {
+      StringRef name;
+      Type type;
+    };
+
+    /// The fields, by offset order
+    SmallVector<Field> fields;
+
+    /// The final MLIR structure type
+    StructType type;
+
+    /// Class name
+    StringRef name;
+
+    /// Reorder fields to make structures more efficient.
+    void optimizeFields()
+    {
+      // TODO: Implement this.
+    }
+
+  public:
+    /// Constructs an empty class with a name
+    ClassInfo(MLIRContext* context, StringRef name) : name(name)
+    {
+      type = StructType::getIdentified(context, name);
+    }
+
+    /// Adds a field to the list before finalisation
+    void addField(StringRef name, Type ty)
+    {
+      assert(!type.isInitialized() && "Can't change type after declaration");
+      fields.push_back({name, ty});
+    }
+
+    /// Finalise structure and set MLIR type
+    void finalize()
+    {
+      // First make sure we have the most optimal placing
+      optimizeFields();
+
+      // Now get the types in order and create the structure
+      SmallVector<Type> types;
+      for (auto field : fields)
+        types.push_back(field.type);
+      auto set = type.setBody(types, /*packed*/ false);
+      // This really shouldn't fail
+      assert(mlir::succeeded(set) && "Error setting fields to class");
+    }
+
+    /// FIXME: Find better map key than opaque pointer types
```

Comment:
Are the MLIR struct types deduplicated?  If I have two classes with the same layout, do they end up aliasing here?

Reply:
There are no MLIR struct types, these are `mlir::LLVM::LLVMStructType` which are uniquely represented by the list of types if unnamed and by their names if named. Their opaque pointer seems like a valid candidate for the key, but not `StructType` which is just a wrapping pointer to the actual implementation.